### PR TITLE
New client for new server/arduino configuration

### PR DESCRIPTION
**Pull request by mgalardini**
# New eVOLVER client script to handle new server/arduino code

Changes proposed in this pull request:
- Merged `eVOLVER_module.py` and `main_eVOLVER.py` in a single file (makes things slightly easier to code, and the user is not supposed to mess with it anyway)
- Custom functions are called when a broadcast is received; this creates the need to asynchronously check if calibrations have been received and if we want to use a blank for OD. On the other hand it removes the need for global variables to be defined everywhere
- Chemostat should be more robust, since the server sends the current configuration with each broadcast; should then automatically restart after all pumps have been stopped with Crtl-C

NOTE: still needs to be tested properly, but open for review if anything looks weird or could be improved

# Checks
- [x] Updated documentation.
- [x] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [x] Follows the [Google Style Guide](https://github.com/google/styleguide).

### CODE REVIEW

**Comment by mgalardini** [comment]
I've just pushed the most recent version of the new code. Unfortunately I couldn't completely finish it, but should be almost there. Here;s a list of issues still open:

- chemostat mode is still not properly tested because the recurring commands are not properly handled by either the arduino or the server (very unclear what is going on there); to be honest I'm unsure whether recurring commands are handled properly for stir and temperature too...
- if the script is paused (Ctrl-C) and restarted after more than ~40 seconds, multiple broadcasts will happen at the same time. I tried to add a semaphore in the `EvolverNameSpace` object but it didn't seem to work, so I removed it for now. there should be an easy fix for this

Happy to review further changes to the code, I'll be traveling next week...

**Comment by mgalardini** [comment]
These now works when together with https://github.com/FYNCH-BIO/evolver/pull/39 and https://github.com/FYNCH-BIO/evolver-arduino/pull/7

I still have to implement a way to avoid multiple broadcast events to overlap, but can be solved in next round of tests

File: experiment/template/eVOLVER.py [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+        print("Reconnected to eVOLVER as client")
+        logger.info("reconnected to eVOLVER as client")
+
+    def on_broadcast(self, data):
+        logger.debug('broadcast received')
+        elapsed_time = round((time.time() - self.start_time) / 3600, 4)
+        logger.debug('elapsed time: %.4f hours' % elapsed_time)
+        print("{0}: {1} Hours".format(EXP_NAME, elapsed_time))
+        # are the calibrations in yet?
+        if self.check_for_calibrations():
```
Comment (zheins) [inline experiment/template/eVOLVER.py]: Instead of doing a bunch of nested if/else logic with increasing indentations, can you just do the check and `return` after logging the error if you don't want to continue? It gets hard to follow when you're 4 nested logic statements in.

File: experiment/template/eVOLVER.py [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+            delta_t = np.abs(temps - temp_data).max()
+            if delta_t > 0.2:
+                logger.info('actual temperature doesn\'t match configuration '
+                            '(yet? max deltaT is %.2f)' % delta_t)
+                logger.debug('temperature config: %s' % temps)
+                logger.debug('actual temperatures: %s' % temp_data)
+
+        # add a new field in the data dictionary
+        data['transformed'] = {}
+        data['transformed']['od_90'] = od_data
```
Comment (zheins) [inline experiment/template/eVOLVER.py]: We should eventually make this generalized (od_135, other experimental params). This is good for now though.

File: experiment/template/eVOLVER.py [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+        data['transformed']['temp'] = temp_data
+        return data
+
+    def update_stir_rate(self, stir_rates):
```
Comment (zheins) [inline experiment/template/eVOLVER.py]: Same comment about generalizing all of this. This is good for updating to handle the new stuff though - let's properly think through how dpu should be generalized later.

File: experiment/template/eVOLVER.py [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+                file_path = os.path.join(EXP_DIR, 'temp_config', file_name)
+                text_file = open(file_path, "w")
+                text_file.write("Experiment: {0} vial {1}, {2}\n".format(
+                    EXP_NAME, x, time.strftime("%c")))
+                #initialize based on custom_script.choose_setup()
+                text_file.write("0,{0}\n".format(TEMP_INITIAL[x]))
+                text_file.close()
+
+                if OPERATION_MODE == 'turbidostat':
+                    # make pump log file
```
Comment (zheins) [inline experiment/template/eVOLVER.py]: Can you make a function do this directory creation stuff? Lots of duplicated code.

File: experiment/template/eVOLVER.py [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+                elif OPERATION_MODE == 'chemostat':
+                    #make chemostat file
+                    file_name =  "vial{0}_chemoconfig.txt".format(x)
+                    file_path = os.path.join(EXP_DIR, 'chemo_config', file_name)
+                    text_file = open(file_path, "w")
+                    text_file.write("0,0,0\n") #header
+                    text_file.write("0,0,0\n") #initialize to 0
+                    text_file.close()
+
+                else:
```
Comment (zheins) [inline experiment/template/eVOLVER.py]: You should probably just always make the pump log. If no mode is specified, don't make those other directories. Maybe let's discuss this with @cpmancuso. You also could just make all of the files even if they aren't used in all cases, which gives some flexibility if the user codes something a little weird, and would clean this area of the code up.

File: experiment/template/eVOLVER.py [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+            x = loaded_var
+            start_time = x[0]
+            self.OD_initial = x[1]
+
+        # copy current custom script to txt file
+        backup_filename = '{0}_{1}.txt'.format(EXP_NAME,
+                                            time.strftime('%y%m%d_%H%M'))
+        shutil.copy('custom_script.py', os.path.join(EXP_DIR,
+                                                    backup_filename))
+        logger.info('saved a copy of current custom_script.py as %s' %
```
Comment (zheins) [inline experiment/template/eVOLVER.py]: 👍  Great idea

File: experiment/template/eVOLVER.py (hunk 4 above)
Comment (mgalardini) [reply experiment/template/eVOLVER.py]: Absolutely!

File: experiment/template/eVOLVER.py (hunk 5 above)
Comment (mgalardini) [reply experiment/template/eVOLVER.py]: Yeah, that makes sense

File: experiment/template/eVOLVER.py (hunk 6 above)
Comment (mgalardini) [reply experiment/template/eVOLVER.py]: It's from our own @cpmancuso!

**Comment by mgalardini** [comment]
Just committed a bunch of changes to address your comments; let me know if anything else can be changed for now!

**Comment by zheins** [comment]
Could you resolve the conflicts and squash the commits into one?

**Comment by mgalardini** [comment]
The only way to merge this PR is to keep the `eVOLVER_module.py` file and remove it after the merge, because computers...

**Comment by mgalardini** [comment]
Never mind, now it can be merged correctly!
